File: frontend/src/services/planService.js
import api from './api';

const planService = {
  async getPlans()          { return (await api.get('/plans')).data; },
  async getPlan(id)         { return (await api.get(`/plans/${id}`)).data; },
  async createPlan(data)    { return (await api.post('/plans', data)).data; },
  async updatePlan(id, data){ return (await api.put(`/plans/${id}`, data)).data; },
  async deletePlan(id)      { await api.delete(`/plans/${id}`); },

  async assignPlan(planId)   { return (await api.post('/user-plan', { plan_id: planId })).data; },
  async getActivePlan()      { return (await api.get('/user-plan/active')).data; },
  async getPlanHistory()     { return (await api.get('/user-plan/history')).data; },
};

export default planService;

File: backend/app/services/plan_service.py
from typing import List
from fastapi import HTTPException, status
from sqlalchemy.orm import Session
from app.repositories.plan_repository import PlanRepository
from app.schemas.plan import PlanCreate, PlanUpdate, PlanOut


class PlanService:
    def __init__(self, db: Session):
        self.repo = PlanRepository(db)

    def list_plans(self, active_only: bool = True) -> List[PlanOut]:
        plans = self.repo.get_active_plans() if active_only else self.repo.get_all()
        return [PlanOut.model_validate(p) for p in plans]

    def get_plan(self, plan_id) -> PlanOut:
        plan = self.repo.get(plan_id)
        if not plan:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Plan not found")
        return PlanOut.model_validate(plan)

    def create_plan(self, data: PlanCreate) -> PlanOut:
        plan = self.repo.create(**data.model_dump())
        return PlanOut.model_validate(plan)

    def update_plan(self, plan_id, data: PlanUpdate) -> PlanOut:
        plan = self.repo.get(plan_id)
        if not plan:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Plan not found")
        updates = {k: v for k, v in data.model_dump().items() if v is not None}
        plan = self.repo.update(plan, **updates)
        return PlanOut.model_validate(plan)

    def delete_plan(self, plan_id) -> None:
        plan = self.repo.get(plan_id)
        if not plan:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Plan not found")
        self.repo.delete(plan)
